Failed tests (10964): TestSendToMultipleClients

The following test appears to have failed:

[#10964](https://circleci.com/gh/cockroachdb/cockroach/10964):

```
W0111 11:11:08.124474 640 rpc/send.go:198  Heartbeat.Ping: error reply: rpc/send.go:247: rpc to Heartbeat.Ping: client not ready after 1s
W0111 11:11:08.192254 640 rpc/server.go:309  rpc: write response failed: write tcp 127.0.0.1:46030->127.0.0.1:39558: use of closed network connection
W0111 11:11:08.235612 640 rpc/client.go:300  unexpected EOF
W0111 11:11:08.375586 640 rpc/client.go:300  unexpected EOF
W0111 11:11:08.393557 640 rpc/client.go:300  unexpected EOF
--- FAIL: TestSendToMultipleClients (1.81s)
	send_test.go:99: failed to send RPC: too many errors encountered (2 of 4 total): rpc/send.go:247: rpc to Heartbeat.Ping: client not ready after 1s
=== RUN   TestRetryableError
W0111 11:11:08.539245 640 rpc/client.go:308  connection is shut down
W0111 11:11:08.542399 640 rpc/client.go:300  dial tcp 127.0.0.1:32952: getsockopt: connection refused
W0111 11:11:08.544120 640 rpc/client.go:300  dial tcp 127.0.0.1:32952: getsockopt: connection refused
W0111 11:11:08.545856 640 rpc/client.go:300  dial tcp 127.0.0.1:32952: getsockopt: connection refused
W0111 11:11:08.547218 640 rpc/client.go:300  dial tcp 127.0.0.1:32952: getsockopt: connection refused
W0111 11:11:08.548780 640 rpc/client.go:300  dial tcp 127.0.0.1:32952: getsockopt: connection refused
W0111 11:11:08.550291 640 rpc/client.go:300  dial tcp 127.0.0.1:32952: getsockopt: connection refused
W0111 11:11:08.552042 640 rpc/client.go:300  dial tcp 127.0.0.1:32952: getsockopt: connection refused
--
	<autogenerated>:28: prior leak detected, leaktest disabled
=== RUN   TestUnregisteredMethod
W0111 11:11:15.213244 640 rpc/send.go:198  Foo.Bar: error reply: rpc/server.go:247: rpc: couldn't find method: Foo.Bar
--- PASS: TestUnregisteredMethod (0.19s)
	<autogenerated>:28: prior leak detected, leaktest disabled
FAIL
FAIL	github.com/cockroachdb/cockroach/rpc	10.685s
=== RUN   TestAll
I0111 11:11:19.152395 706 rpc/codec/rpc_test.go:48  Arith.Add(1, 2): 3
I0111 11:11:19.155438 706 rpc/codec/rpc_test.go:48  Arith.Add(1, 2): 3
--- PASS: TestAll (0.01s)
PASS
ok  	github.com/cockroachdb/cockroach/rpc/codec	1.399s
?   	github.com/cockroachdb/cockroach/rpc/codec/message	[no test files]
?   	github.com/cockroachdb/cockroach/rpc/codec/wire	[no test files]
=== RUN   TestGetCertificateUser
--- PASS: TestGetCertificateUser (0.00s)

```
Please assign, take a look and update the issue accordingly.

--- a/rpc/client.go
+++ b/rpc/client.go
@@ -45,9 +45,8 @@ const (
 )
 
 var (
-	clientMu          sync.Mutex         // Protects access to the client cache.
-	clients           map[string]*Client // Cache of RPC clients.
-	heartbeatInterval = defaultHeartbeatInterval
+	clientMu sync.Mutex         // Protects access to the client cache.
+	clients  map[string]*Client // Cache of RPC clients.
 	// TODO(tschottdorf) err{Closed,Unstarted} are candidates for NodeUnavailableError.
 	errClosed    = errors.New("client is closed")
 	errUnstarted = errors.New("not started yet")
@@ -91,6 +90,8 @@ type Client struct {
 	clock             *hlc.Clock
 	remoteClocks      *RemoteClockMonitor
 	remoteOffset      RemoteOffset
+	heartbeatInterval time.Duration
+	heartbeatTimeout  time.Duration
 }
 
 // NewClient returns a client RPC stub for the specified address
@@ -131,6 +132,15 @@ func NewClient(addr net.Addr, context *Context) *Client {
 		disableReconnects: context.DisableReconnects,
 		clock:             context.localClock,
 		remoteClocks:      context.RemoteClocks,
+		heartbeatInterval: context.heartbeatInterval,
+		heartbeatTimeout:  context.heartbeatTimeout,
+	}
+
+	if c.heartbeatInterval == 0 {
+		c.heartbeatInterval = defaultHeartbeatInterval
+	}
+	if c.heartbeatTimeout == 0 {
+		c.heartbeatTimeout = 2 * c.heartbeatInterval
 	}
 
 	c.healthy.Store(make(chan struct{}))
@@ -248,6 +258,18 @@ func (c *Client) close() {
 	}
 }
 
+func (c *Client) maybeClose(extCloser <-chan struct{}) bool {
+	select {
+	case <-c.closer:
+		return true
+	case <-extCloser:
+		c.close()
+		return true
+	default:
+		return false
+	}
+}
+
 // runHeartbeat sends periodic heartbeats to client, marking the client healthy
 // or unhealthy and reconnecting appropriately until either the Client or the
 // supplied channel is closed.
@@ -280,13 +302,8 @@ func (c *Client) runHeartbeat(retryOpts retry.Options) {
 	var err = errUnstarted // initial condition
 	for {
 		for r := retry.Start(retryOpts); r.Next(); {
-			select {
-			case <-c.closer:
+			if c.maybeClose(retryOpts.Closer) {
 				return
-			case <-retryOpts.Closer:
-				c.close()
-				return
-			default:
 			}
 
 			// Reconnect on failure.
@@ -306,6 +323,9 @@ func (c *Client) runHeartbeat(retryOpts retry.Options) {
 			if err = c.heartbeat(retryOpts.Closer); err != nil {
 				setUnhealthy()
 				log.Warning(err)
+				if c.maybeClose(retryOpts.Closer) {
+					return
+				}
 				continue
 			}
 
@@ -321,7 +341,7 @@ func (c *Client) runHeartbeat(retryOpts retry.Options) {
 		case <-retryOpts.Closer:
 			c.close()
 			return
-		case <-time.After(heartbeatInterval):
+		case <-time.After(c.heartbeatInterval):
 			// TODO(tamird): Perhaps retry more aggressively when the client is unhealthy.
 		}
 	}
@@ -357,8 +377,8 @@ func (c *Client) heartbeat(closer <-chan struct{}) error {
 		if err := call.Error; err != nil {
 			return err
 		}
-	case <-time.After(2 * heartbeatInterval):
-		return util.Errorf("heartbeat timed out after %s", 2*heartbeatInterval)
+	case <-time.After(c.heartbeatTimeout):
+		return util.Errorf("heartbeat timed out after %s", c.heartbeatTimeout)
 	}
 
 	receiveTime := c.clock.PhysicalNow()

--- a/rpc/client_test.go
+++ b/rpc/client_test.go
@@ -202,7 +202,7 @@ func TestDelayedOffsetMeasurement(t *testing.T) {
 	context.RemoteClocks.mu.Unlock()
 }
 
-func TestFailedOffestMeasurement(t *testing.T) {
+func TestFailedOffsetMeasurement(t *testing.T) {
 	defer leaktest.AfterTest(t)
 
 	stopper := stop.NewStopper()
@@ -227,9 +227,11 @@ func TestFailedOffestMeasurement(t *testing.T) {
 	clientManual := hlc.NewManualClock(0)
 	clientClock := hlc.NewClock(clientManual.UnixNano)
 	context := newNodeTestContext(clientClock, stopper)
+	context.heartbeatTimeout = 20 * context.heartbeatInterval
 	c := NewClient(ln.Addr(), context)
 	heartbeat.ready <- struct{}{} // Allow one heartbeat for initialization.
 	<-c.Healthy()
+
 	// Synchronously wait on missing the next heartbeat.
 	if err := util.IsTrueWithin(func() bool {
 		select {
@@ -238,7 +240,7 @@ func TestFailedOffestMeasurement(t *testing.T) {
 		default:
 			return true
 		}
-	}, heartbeatInterval*10); err != nil {
+	}, context.heartbeatTimeout*10); err != nil {
 		t.Fatal(err)
 	}
 	if !proto.Equal(&c.remoteOffset, &RemoteOffset{}) {

--- a/rpc/clock_offset.go
+++ b/rpc/clock_offset.go
@@ -34,7 +34,7 @@ var (
 )
 
 func init() {
-	monitorInterval = heartbeatInterval * 10
+	monitorInterval = defaultHeartbeatInterval * 10
 }
 
 // RemoteClockMonitor keeps track of the most recent measurements of remote

--- a/rpc/main_test.go
+++ b/rpc/main_test.go
@@ -38,12 +38,10 @@ func init() {
 //go:generate ../util/leaktest/add-leaktest.sh *_test.go
 
 func TestMain(m *testing.M) {
-	defer func(hbInterval time.Duration, retryOpts retry.Options) {
-		heartbeatInterval = hbInterval
+	defer func(retryOpts retry.Options) {
 		clientRetryOptions = retryOpts
-	}(heartbeatInterval, clientRetryOptions)
+	}(clientRetryOptions)
 
-	heartbeatInterval = 10 * time.Millisecond
 	clientRetryOptions = retry.Options{
 		InitialBackoff: 1 * time.Millisecond,
 		MaxBackoff:     1 * time.Millisecond,
@@ -58,7 +56,10 @@ func newNodeTestContext(clock *hlc.Clock, stopper *stop.Stopper) *Context {
 	if clock == nil {
 		clock = hlc.NewClock(hlc.UnixNano)
 	}
-	return NewContext(testutils.NewNodeTestBaseContext(), clock, stopper)
+	ctx := NewContext(testutils.NewNodeTestBaseContext(), clock, stopper)
+	ctx.heartbeatInterval = 10 * time.Millisecond
+	ctx.heartbeatTimeout = 2 * defaultHeartbeatInterval
+	return ctx
 }
 
 func newTestServer(t *testing.T, ctx *Context, manual bool) (*Server, net.Listener) {

--- a/rpc/rpc.go
+++ b/rpc/rpc.go
@@ -19,6 +19,9 @@ type Context struct {
 	DisableCache      bool // Disable client cache when calling NewClient()
 	DisableReconnects bool // Disable client reconnects
 	HealthWait        time.Duration
+
+	heartbeatInterval time.Duration
+	heartbeatTimeout  time.Duration
 }
 
 // NewContext creates an rpc Context with the supplied values.

--- a/rpc/send_test.go
+++ b/rpc/send_test.go
@@ -59,7 +59,7 @@ func TestSendToOneClient(t *testing.T) {
 		N:               1,
 		Ordering:        OrderStable,
 		SendNextTimeout: 1 * time.Second,
-		Timeout:         1 * time.Second,
+		Timeout:         10 * time.Second,
 	}
 	replies, err := sendPing(opts, []net.Addr{ln.Addr()}, ctx)
 	if err != nil {
@@ -92,7 +92,7 @@ func TestSendToMultipleClients(t *testing.T) {
 			N:               n,
 			Ordering:        OrderStable,
 			SendNextTimeout: 1 * time.Second,
-			Timeout:         1 * time.Second,
+			Timeout:         10 * time.Second,
 		}
 		replies, err := sendPing(opts, addrs, nodeContext)
 		if err != nil {
@@ -113,6 +113,7 @@ func TestRetryableError(t *testing.T) {
 	defer stopper.Stop()
 
 	nodeContext := newNodeTestContext(nil, stopper)
+	nodeContext.heartbeatTimeout = 10 * nodeContext.heartbeatInterval
 	_, ln := newTestServer(t, nodeContext, false)
 
 	c := NewClient(ln.Addr(), nodeContext)
@@ -133,8 +134,8 @@ func TestRetryableError(t *testing.T) {
 	opts := Options{
 		N:               1,
 		Ordering:        OrderStable,
-		SendNextTimeout: 1 * time.Second,
-		Timeout:         1 * time.Second,
+		SendNextTimeout: 100 * time.Millisecond,
+		Timeout:         100 * time.Millisecond,
 	}
 	if _, err := sendPing(opts, []net.Addr{ln.Addr()}, nodeContext); err != nil {
 		retryErr, ok := err.(retry.Retryable)
@@ -172,7 +173,7 @@ func TestUnretryableError(t *testing.T) {
 		N:               1,
 		Ordering:        OrderStable,
 		SendNextTimeout: 1 * time.Second,
-		Timeout:         5 * time.Second,
+		Timeout:         10 * time.Second,
 	}
 	getArgs := func(addr net.Addr) proto.Message {
 		return &roachpb.Span{}
@@ -241,28 +242,27 @@ func TestClientNotReady(t *testing.T) {
 	// Send the RPC again with no timeout.
 	opts.SendNextTimeout = 0
 	opts.Timeout = 0
-	c := make(chan struct{})
+	c := make(chan error)
 	go func() {
 		if _, err := sendPing(opts, []net.Addr{ln.Addr()}, nodeContext); err == nil {
-			t.Fatalf("expected error when client is closed")
+			c <- util.Errorf("expected error when client is closed")
 		} else if !strings.Contains(err.Error(), "failed as client connection was closed") {
-			t.Fatal(err)
+			c <- err
 		}
 		close(c)
 	}()
+
 	select {
 	case <-c:
 		t.Fatalf("Unexpected end of rpc call")
 	case <-time.After(1 * time.Millisecond):
 	}
 
-	// Grab the client for our invalid address and close it. This will
-	// cause the blocked ping RPC to finish.
+	// Grab the client for our invalid address and close it. This will cause the
+	// blocked ping RPC to finish.
 	NewClient(ln.Addr(), nodeContext).Close()
-	select {
-	case <-c:
-	case <-time.After(100 * time.Millisecond):
-		t.Fatalf("RPC call failed to return")
+	if err := <-c; err != nil {
+		t.Fatal(err)
 	}
 }
 
@@ -323,7 +323,7 @@ func TestComplexScenarios(t *testing.T) {
 			N:               test.numRequests,
 			Ordering:        OrderStable,
 			SendNextTimeout: 1 * time.Second,
-			Timeout:         1 * time.Second,
+			Timeout:         10 * time.Second,
 		}
 		getArgs := func(addr net.Addr) proto.Message {
 			return &PingRequest{}

